app/models/solution/tag.rb

```diff
@@ -0,0 +1,21 @@
+class Solution::Tag < ApplicationRecord
+  extend Mandate::Memoize
+
+  belongs_to :solution
+  belongs_to :exercise
+  belongs_to :user
+
+  before_validation on: :create do
+    self.exercise = solution.exercise unless exercise
+    self.user = solution.user unless user
```

Comment:
Would probably be better to set `exercise_id` and `user_id` here to avoid loading the records from the db. (Thinking if we do batch things this could add a lot of extra DB calls)

Reply:
Makes sense. We should probably do that in other places too.